criu/cr-restore.c

```diff
@@ -3201,6 +3201,35 @@ static void rst_reloc_creds(struct thread_restore_args *thread_args,
 	thread_args->creds_args = args;
 }
 
+static bool groups_match(gid_t* groups, int n_groups)
+{
+	int n, len;
+	bool ret;
+	gid_t* gids;
+
+	n = getgroups(0, NULL);
+	if (n != n_groups)
+		return false;
+	if (n == 0)
+		return true;
+
+	len = n * sizeof(gid_t);
+	gids = xmalloc(len);
+	if (gids == NULL)
+		return false;
+
+	n = getgroups(n, gids);
+	if (n == -1) {
+		pr_perror("Failed to get supplementary groups");
+		ret = false;
+	} else {
+		ret = !memcmp(gids, groups, len);
```

Comment:
memcp will work only if both arrays are sorted. Are they sorted? if so please add a comment here.

Reply:
Good catch. It appears that the `groups` variable in this function is set from [here](https://github.com/checkpoint-restore/criu/blob/criu-dev/criu/pie/parasite.c#L277) with `sys_getgroups`, so the order *should* be the same. However, on the off-chance that these two arrays are not in the same order, the restoring will revert to its default process as if this `groups_match` function wasn't added. Let me know if you'd prefer some form of sorting before calling the `memcpy`!

Reply:
The right answer is that the Linux getgroups returns groups in the sorted order: https://github.com/torvalds/linux/blob/master/kernel/groups.c#L209.

There is no guarantee that this will not be changed in the future, but I think it will not be done soon.

I was lazy to read the kernel code before writing the previous comment. Pls, add a comment into the code.